extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxRecorder.java

```diff
@@ -101,16 +109,22 @@ public void handle(Promise<Object> event) {
                                 }
                             }, invoker.isOrdered(), null);
                         } else {
-                            try {
-                                invoker.invoke(m);
-                            } catch (Exception e) {
-                                if (m.replyAddress() == null) {
-                                    // No reply handler
-                                    throw wrapIfNecessary(e);
-                                } else {
-                                    m.fail(ConsumeEvent.FAILURE_CODE, e.toString());
+                            // Use the capture context to invoke the method on the event loop.
+                            context.runOnContext(new Handler<Void>() {
```

Comment:
instead execute `runOnContext` when the consumer is created, i.e

context.runOnContext(v -> {
  MessageConsumer consumer = eventBus.consumer(address);
});

the dispatch will then be done by vertx on this context precisely

Reply:
I tried like this, but we need to wait for the completion which leads to interesting races. I will have another look.